### PR TITLE
REF: Add related impls to initial items to move

### PR DESCRIPTION
**Pull request by dima74**
E.g. when moving some struct, all its impls will be added to initial items to move

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt [hunk 1]
```diff
@@ -84,3 +85,28 @@ class RsMoveTopLevelItemsHandler : MoveHandlerDelegate() {
         }
     }
 }
+
+fun collectRelatedImplItems(containingMod: RsMod, items: List<RsItemElement>): List<RsImplItem> {
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt:89]:
```suggestion
private fun collectRelatedImplItems(containingMod: RsMod, items: List<RsItemElement>): List<RsImplItem> {
```

File: src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt [hunk 2]
```diff
@@ -84,3 +85,28 @@ class RsMoveTopLevelItemsHandler : MoveHandlerDelegate() {
...
+    // * impl ... for Foo { ... }
+    // * Maybe also `impl From<Foo> for Bar { ... }`?
+    //   if `Bar` belongs to same crate (but to different module from `Foo`)
+    //
+    // For trait `Foo` we should collect:
+    // * impl Foo for ... { ... }
+    return groupImplsByStructOrTrait(containingMod, items).values.flatten()
+}
+
+fun groupImplsByStructOrTrait(containingMod: RsMod, items: List<RsItemElement>): Map<RsItemElement, List<RsImplItem>> {
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt:101]:
```suggestion
private fun groupImplsByStructOrTrait(containingMod: RsMod, items: List<RsItemElement>): Map<RsItemElement, List<RsImplItem>> {
```

File: src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt [hunk 3]
```diff
@@ -84,3 +85,28 @@ class RsMoveTopLevelItemsHandler : MoveHandlerDelegate() {
...
+    // * impl ... for Foo { ... }
+    // * Maybe also `impl From<Foo> for Bar { ... }`?
+    //   if `Bar` belongs to same crate (but to different module from `Foo`)
+    //
+    // For trait `Foo` we should collect:
+    // * impl Foo for ... { ... }
+    return groupImplsByStructOrTrait(containingMod, items).values.flatten()
+}
+
+private fun groupImplsByStructOrTrait(containingMod: RsMod, items: List<RsItemElement>): Map<RsItemElement, List<RsImplItem>> {
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt:101]: [minor] Probably  `items: Set<RsItemElement>` is preferable as we only do lookups

File: src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt [hunk 4]
```diff
@@ -84,3 +85,28 @@ class RsMoveTopLevelItemsHandler : MoveHandlerDelegate() {
...
+    return containingMod
+        .childrenOfType<RsImplItem>()
+        .mapNotNull { impl ->
+            val struct = (impl.typeReference?.type as? TyAdt)?.item
+            val trait = impl.traitRef?.path?.reference?.resolve() as? RsTraitItem
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt:106]: Let's use `as? RsItemElement` here to get rid of `as RsItemElement?` cast below

File: src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt [hunk 5]
```diff
@@ -84,3 +85,28 @@ class RsMoveTopLevelItemsHandler : MoveHandlerDelegate() {
...
+            val relatedItem = struct?.takeIf { items.contains(it) } as RsItemElement?
+                ?: trait?.takeIf { items.contains(it) }
+            if (relatedItem != null) relatedItem to impl else null
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt:109]: [minor] Probably more clear version
```suggestion
                ?: return@mapNotNull null
            relatedItem to impl
```

**Comment by dima74** [comment]
bors r=ortem

**Pull request merged**